### ISSUE TITLE
Allow slice-wise and voxel-wise regressors in GLMs and Maskers

### ISSUE THREAD

**Issue by tsalo** [nilearn/nilearn#3418]
At the moment, I don't think Nilearn supports slice-wise (e.g., RETROICOR confounds) or voxel-wise regressors (e.g., spatially-varying HRFs). I think this could be useful, though.

### Benefits to the change
This would make it easier to perform certain types of denoising, or use certain tools (e.g., RETROICOR), with Nilearn.

### Pseudocode for the new behavior, if applicable

Perhaps something like this?

```python
masker.transform(
    data_img,
    confounds=confounds_df,  # regular confounds in a DataFrame
    voxelwise_confounds=[confounds_img1, confounds_img2],
    slicewise_confounds=slicewise_confounds_df,  # a DataFrame, maybe with columns indicating the slice
)
```

**Comment by bthirion** [comment]
To be frank, I'm a bit afraid of this beast. 
This would probably require a novel masker object that essentially implements several NiftiMaskers  in parallel. Otherwise, the NiftiMasker API will explode.

**Comment by tsalo** [comment]
I agree that it's a pretty difficult extension. It may be that it is not a common enough use-case to be worth the effort, but I _do_ think it falls within nilearn's scope. Perhaps we could solicit user opinions? Or just label it as a long-term roadmap issue, with no immediate plans to tackle it?

**Comment by bthirion** [comment]
Both suggestions are fine. For the time being, are there references that explicitly report a benefit of such partial models ? I also not that in the case of the GLM, we already hinge on a somewhat similar approach, when we group voxels according to autocorrelation profile.  But this is done internally and is not meant to be exposed to users so far.

**Comment by tsalo** [comment]
I don't know about any papers that directly compare these kinds of regressors against generic ones, but slice-wise regressors are used in RETROICOR, and I believe some tools, like [rapidtide](https://rapidtide.readthedocs.io/en/latest/index.html), estimate spatially-varying nuisance regressors.

**Comment by emdupre** [comment]
I guess I'm not entirely convinced that this falls within Nilearn's scope ; I'd suggest that both RETROICOR and rapidtide for denoising before subsequent analysis, not for statistical or machine learning analysis itself -- and I think that's an important distinction.

That's not to say that all spatially or time-varying regressors are for denoising only: tools like [GLMsingle](https://github.com/cvnlab/GLMsingle), for example, do use spatially varying HRFs. Nonetheless, I'd argue its philosophy and implementation are a bit too far from the "approachable and versatile analysis" that defines our scope to consider inclusion.

All that to say : I'd argue that before even adding to a roadmap or asking users, there's still a large burden of proof on the idea that these kinds of regressors really are in scope.

**Comment by tsalo** [comment]
>I'd suggest that both RETROICOR and rapidtide for denoising before subsequent analysis, not for statistical or machine learning analysis itself -- and I think that's an important distinction.

Personally, I'd like to see denoising methods like rapidtide run in conjunction with a modeling GLM, but I can understand why that use-case might not be worth the effort.

>I'd argue that before even adding to a roadmap or asking users, there's still a large burden of proof on the idea that these kinds of regressors really are in scope.

I'm happy to close this as "not planned". If users start requesting it en masse (seems unlikely), we can revisit my proposal. Otherwise, we can point users to other tools as needed. Does that sound good?

**Comment by bthirion** [comment]
Yes. Thx for opening the discussion.
